src/modules/content-s3/content-s3.c

```diff
@@ -0,0 +1,300 @@
+        errno = EIO;
+        flux_log_error (h, "error responding to kvs-checkpoint.get request (pack)");
+    }
+    free (data);
+    free (dup);
+    return;
+
+error:
+    if (errstr)
+        printf("\n\n%s\n\n", errstr);
+    if (flux_respond_error (h, msg, errno, errstr) < 0)
+        flux_log_error (h, "error responding to kvs-checkpoint.get request");
+    free (data);
+    free (dup);
+}
+
+/* Handle a kvs-checkpoint.put request from the rank 0 kvs module.
+ * The KVS stores its last root reference here for restart purposes.
+ */
+void checkpoint_put_cb (flux_t *h, flux_msg_handler_t *mh, const flux_msg_t *msg, void *arg)
+{
+    struct content_s3 *ctx = arg;
+    const char *key;
+    const char *value;
+    const char *errstr = NULL;
+
+    if (flux_request_unpack (msg,
+                             NULL,
+                             "{s:s s:s}",
+                             "key",
+                             &key,
+                             "value",
+                             &value) < 0)
+        goto error;
+    if (s3_put (ctx->cfg, key, value, strlen (value), &errstr) < 0)
+        goto error;
+    if (flux_respond (h, msg, NULL) < 0) {
+        errno = EIO;
+        flux_log_error (h, "error responding to kvs-checkpoint.put request (pack)");
+    }
+    return;
+
+error:
+    if (errstr)
+        printf("\n\n%s\n\n", errstr);
+    if (flux_respond_error (h, msg, errno, errstr) < 0)
+        flux_log_error (h, "error responding to kvs-checkpoint.put request");
+}
+
+/* Table of message handler callbacks registered below.
+ * The topic strings in the table consist of <service name>.<method>.
+ */
+static const struct flux_msg_handler_spec htab[] = {
+    { FLUX_MSGTYPE_REQUEST, "content-backing.load",    load_cb, 0 },
+    { FLUX_MSGTYPE_REQUEST, "content-backing.store",   store_cb, 0 },
+    { FLUX_MSGTYPE_REQUEST, "kvs-checkpoint.get", checkpoint_get_cb, 0 },
+    { FLUX_MSGTYPE_REQUEST, "kvs-checkpoint.put", checkpoint_put_cb, 0 },
+    FLUX_MSGHANDLER_TABLE_END,
+};
+
+/* Destroy module context.
+ */
+static void content_s3_destroy (struct content_s3 *ctx)
+{
+    if(ctx) {
+        int saved_errno = errno;
+        flux_msg_handler_delvec (ctx->handlers);
+        free (ctx->cfg);
+        free (ctx);
+        errno = saved_errno;
+    }
+
+    s3_cleanup();
+}
+
+/* Create the s3 context, initalize the connection, and
+ * create the working bucket
+ */
+static struct content_s3 *content_s3_create (flux_t *h)
+{
+    const char *errstr = NULL;
+    struct content_s3 *ctx;
+    const char *tmp;
+
+    if (!(ctx = calloc (1, sizeof (*ctx))))
+        return NULL;
+    ctx->h = h;
+
+    if (!(ctx->hashfun = flux_attr_get (h, "content.hash"))) {
+        flux_log_error (h, "content.hash");
+        goto error;
+    }
+    if (!(tmp = flux_attr_get (h, "content.blob-size-limit"))) {
+        flux_log_error (h, "content.blob-size-limit");
+        goto error;
+    }
+    ctx->blob_size_limit = strtoul (tmp, NULL, 10);
```

Comment:
We decided to remove the blob-size-limit check from all the content backends in the content-files PR so this can be removed.